### PR TITLE
[RISCV] Implement foward inserting save/restore FRM instructions.

### PR DESCRIPTION
**Pull request by yetingk**
Previously, RISCVInsertReadWriteCSR inserted an FRM swap for any value other than 7 and restored the original value right after the vector instruction. This is inefficient if multiple vector instructions use the same rounding mode or if the next vector instruction uses a different explicit rounding mode.

This patch implements a local optimization to solve the above problem. We assume the starting rounding mode of the basic block is "dynamic." When iterating through a basic block and encountering an instruction whose rounding mode is not the same as the current rounding mode, we change the current rounding mode and save the current rounding mode if needed. And we may need to restore FRM when encountering function call, inline asm and some uses of FRM.

The advanced version of this is to perform cross basic block analysis for the starting rounding mode of each basic block.

### CODE REVIEW

**Comment by lukel97** [comment]
> The advanced version of this is to perform cross basic block analysis for the starting rounding mode of each basic block. 

I wonder if there's a way we could abstract the cross block analysis in RISCVInsertVSETVLI?

**Comment by topperc** [comment]
> > The advanced version of this is to perform cross basic block analysis for the starting rounding mode of each basic block.
> 
> I wonder if there's a way we could abstract the cross block analysis in RISCVInsertVSETVLI?

Or the one in VXRM insertion. I think the hard thing with FRM is managing the virtual register needed to hold the previous dynamic value.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 1]
```diff
@@ -55,6 +60,109 @@ char RISCVInsertReadWriteCSR::ID = 0;
...
+    auto getInstructionRM = [](MachineInstr &MI) -> std::optional<unsigned> {
+      int FRMIdx = RISCVII::getFRMOpNum(MI.getDesc());
+      if (FRMIdx >= 0)
+        return MI.getOperand(FRMIdx).getImm();
+
+      if (!MI.hasRegisterImplicitUseOperand(RISCV::FRM))
+        return std::nullopt;
+
+      // FIXME: Return nullopt if the rounding mode of MI is not DYN, like
+      // FADD_S with RTZ.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:106]: Does FADD_S with RTZ have an FRM implicit use? It wasn't supposed to.

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    i64 0, i64 %2)
+  %b = call <vscale x 1 x float> @llvm.riscv.vfadd.nxv1f32.nxv1f32(
+    <vscale x 1 x float> undef,
+    <vscale x 1 x float> %a,
+    <vscale x 1 x float> %a,
+    i64 1, i64 %2)
+  ret <vscale x 1 x float> %b
+}
+
+; Test restoring frm before function call and doing nothing with folling dynamic
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:59]: folling -> following

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  call void @foo()
+  %b = call <vscale x 1 x float> @llvm.riscv.vfadd.nxv1f32.nxv1f32(
+    <vscale x 1 x float> undef,
+    <vscale x 1 x float> %a,
+    <vscale x 1 x float> %a,
+    i64 7, i64 %2)
+  ret <vscale x 1 x float> %b
+}
+
+; Test restoring frm before inline asm and doing nothing with folling dynamic
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:105]: folling -> following

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  call void asm sideeffect "", ""()
+  %b = call <vscale x 1 x float> @llvm.riscv.vfadd.nxv1f32.nxv1f32(
+    <vscale x 1 x float> undef,
+    <vscale x 1 x float> %a,
+    <vscale x 1 x float> %a,
+    i64 7, i64 %2)
+  ret <vscale x 1 x float> %b
+}
+
+; Test restoring frm before reading frmm and doing nothing with folling dynamic
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:134]: frmm -> frm
folling -> following

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 5]
```diff
@@ -0,0 +1,217 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=riscv64 -mattr=+v -verify-machineinstrs -target-abi=lp64d < %s | FileCheck %s
+
+declare <vscale x 1 x float> @llvm.riscv.vfadd.nxv1f32.nxv1f32(
+  <vscale x 1 x float>,
+  <vscale x 1 x float>,
+  <vscale x 1 x float>,
+  i64, i64);
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:8]: No need for `;` at the end of this line.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 6]
```diff
@@ -55,6 +61,107 @@ char RISCVInsertReadWriteCSR::ID = 0;
...
+    std::optional<unsigned> InstrRM = getInstructionRM(MI);
+
+    // Skip if MI does not need FRM.
+    if (!InstrRM.has_value())
+      continue;
+
+    if (InstrRM != RISCVFPRndMode::DYN)
+      LastFRMChanger = &MI;
+
+    if (!MI.readsRegister(RISCV::FRM))
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:118]: I believe this if is redundant as you'd have early-continued above.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 7]
```diff
@@ -55,6 +61,107 @@ char RISCVInsertReadWriteCSR::ID = 0;
...
+      MI.addOperand(MachineOperand::CreateReg(RISCV::FRM, /*IsDef*/ false,
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:119]: Must set Changed here.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 8]
```diff
@@ -55,6 +61,107 @@ char RISCVInsertReadWriteCSR::ID = 0;
...
+    if (InstrRM == CurrentRM)
+      continue;
+
+    if (InstrRM == RISCVFPRndMode::DYN) {
+      if (!SavedFRM.has_value())
+        continue;
+      // SavedFRM not having a value means current FRM has correct rounding
+      // mode.
+      BuildMI(MBB, MI, MI.getDebugLoc(), TII->get(RISCV::WriteFRM))
+          .addReg(*SavedFRM);
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:132]: Must set Changed here.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 9]
```diff
@@ -55,6 +61,107 @@ char RISCVInsertReadWriteCSR::ID = 0;
...
+  for (MachineInstr &MI : MBB) {
+    if (MI.getOpcode() == RISCV::SwapFRMImm ||
+        MI.getOpcode() == RISCV::WriteFRMImm) {
+      CurrentRM = MI.getOperand(0).getImm();
+      SavedFRM = std::nullopt;
+      continue;
+    }
+
+    if (MI.getOpcode() == RISCV::WriteFRM) {
+      CurrentRM = RISCVFPRndMode::DYN;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:80]: I'm not sure here, but I think you may be confusing CurrentRM being DYN and CurrentRM being unknown.  I don't know if it matters, but I think if I'm reading this right that DYN is supposed to be a "don't care" value, not an "unknown" value.

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    <vscale x 1 x float> %a,
+    <vscale x 1 x float> %a,
+    i64 1, i64 %2)
+  ret <vscale x 1 x float> %b
+}
+
+; Test restoring frm before function call and doing nothing with following
+; dynamic rounding mode operations.
+declare void @foo()
+define <vscale x 1 x float> @test3(<vscale x 1 x float> %0, <vscale x 1 x float> %1, i64 %2) nounwind {
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:62]: Can you add test cases for the following sub-cases?
* Call only, no operations which modify FRM
* Call before FRM modification
* Call after FRM modification
* Call after DYN use
* Call before DYN use

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+; dynamic rounding mode operations.
+declare i32 @llvm.get.rounding()
+define <vscale x 1 x float> @test5(<vscale x 1 x float> %0, <vscale x 1 x float> %1, i64 %2, ptr %p) nounwind {
+; CHECK-LABEL: test5:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    vsetvli zero, a0, e32, mf2, ta, ma
+; CHECK-NEXT:    fsrmi a0, 0
+; CHECK-NEXT:    vfadd.vv v8, v8, v9
+; CHECK-NEXT:    fsrm a0
+; CHECK-NEXT:    frrm a0
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:144]: We should be able to elide the read here.  Not sure if worth implementing, just noting opportunity.

File: llvm/test/CodeGen/RISCV/rvv/frm-insert.ll [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+declare void @llvm.set.rounding(i32)
+define <vscale x 1 x float> @test6(<vscale x 1 x float> %0, <vscale x 1 x float> %1, i64 %2) nounwind {
+; CHECK-LABEL: test6:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    fsrmi 4
+; CHECK-NEXT:    vsetvli zero, a0, e32, mf2, ta, ma
+; CHECK-NEXT:    vfadd.vv v8, v8, v9
+; CHECK-NEXT:    vfadd.vv v8, v8, v8
+; CHECK-NEXT:    ret
+entry:
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/frm-insert.ll:179]: Please split this into two tests: one for following known same, one for following DYN.

Also please add coverage for the case where the immediate next is known *different* constant.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 13]
```diff
@@ -55,6 +61,82 @@ char RISCVInsertReadWriteCSR::ID = 0;
 INITIALIZE_PASS(RISCVInsertReadWriteCSR, DEBUG_TYPE,
                 RISCV_INSERT_READ_WRITE_CSR_NAME, false, false)
 
+// TODO: Use more accurate rounding mode at the start of MBB.
+bool RISCVInsertReadWriteCSR::emitWriteRoundingModeOpt(MachineBasicBlock &MBB) {
+  bool Changed = false;
+  MachineInstr *LastFRMChanger = nullptr;
+  unsigned CurrentRM = RISCVFPRndMode::DYN;
+  std::optional<Register> SavedFRM;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:69]: I don't think we need std::optional. We can use `SavedFRM.isValid()` to track validity. And we can use `SavedFRM = Register()` to make it invalid when we need to.

File: llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp [hunk 14]
```diff
@@ -99,8 +181,12 @@ bool RISCVInsertReadWriteCSR::runOnMachineFunction(MachineFunction &MF) {
 
   bool Changed = false;
 
-  for (MachineBasicBlock &MBB : MF)
-    Changed |= emitWriteRoundingMode(MBB);
+  for (MachineBasicBlock &MBB : MF) {
+    if (DisableFRMInsertOpt)
+      Changed |= emitWriteRoundingMode(MBB);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInsertReadWriteCSR.cpp:186]: No test coverage for the unoptimized version.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**